cases/src/org/commcare/cases/util/CasePurgeFilter.java

```diff
@@ -132,61 +126,91 @@ private void setIdsToRemoveNew(IStorageUtilityIndexed<Case> caseStorage, Vector<
-                }
-                //add this to the process stack
-                //TODO: We're going to walk stuff redundantly this
-                //way. Can we optimize?
-                //If parent is already live, does
-                //that necessarily denote that its ancestors are alive?
-                toProcess.addElement(edge.i);
+    private void markRelevant(DAG<String, int[], String> g) {
+        walk(g,true, STATUS_RELEVANT, STATUS_RELEVANT);
+        walk(g,false, STATUS_RELEVANT, STATUS_RELEVANT, CaseIndex.RELATIONSHIP_EXTENSION);
+    }
+
+    private void markAvailable(DAG<String, int[], String> g) {
+        for(Enumeration<String> e = g.getIndices(); e.hasMoreElements();) {
+            String index = e.nextElement();
+            int[] node = g.getNode(index);
+            if(is(node, STATUS_OPEN | STATUS_RELEVANT) &&
+                    !hasOutgoingExtension(g,index)) {
+                node[0] |= STATUS_AVAILABLE;
+            }
+        }
+        walk(g,false, STATUS_AVAILABLE, STATUS_AVAILABLE, CaseIndex.RELATIONSHIP_EXTENSION);
+    }
+
+    private void markLive(DAG<String, int[], String> g) {
+        for(Enumeration<String> e = g.getIndices(); e.hasMoreElements();) {
+            String index = e.nextElement();
+            int[] node = g.getNode(index);
+            if(is(node, STATUS_OWNED | STATUS_RELEVANT | STATUS_AVAILABLE)) {
+                node[0] |= STATUS_ALIVE;
+            }
+        }
+
+        walk(g, true, STATUS_ALIVE, STATUS_ALIVE);
+        walk(g, false, STATUS_ALIVE, STATUS_ALIVE, CaseIndex.RELATIONSHIP_EXTENSION);
+    }
+
+    private boolean hasOutgoingExtension(DAG<String, int[], String> g, String index) {
+        for(Edge<String, String> edge : g.getChildren(index)) {
+            if(edge.e.equals(CaseIndex.RELATIONSHIP_EXTENSION)) {
+                return true;
             }
         }
+        return false;
+    }
 
-        //Pass 2: Start at Sinks (nodes which only receive edges) and walk down the
-        //graphs, marking necessary nodes as alive
-        toProcess = g.getSinks();
+    private void walk(DAG<String, int[], String> g, boolean direction, int mask, int mark) {
+        walk(g, direction, mask, mark, null);
+    }
+
+    private void walk(DAG<String, int[], String> g, boolean direction, int mask, int mark, String relationship) {
+        Stack<String> toProcess = direction ? g.getSources() : g.getSinks();
         while (!toProcess.isEmpty()) {
             //current node
             String index = toProcess.pop();
-            String[] node = g.getNode(index);
-
-            //Walk all indexing nodes.
-            for (Edge<String, String> edge : g.getParents(index)) {
-                if (g.getNode(edge.i)[0].equals(STATUS_ABANDONED) &&
-                        edge.e.equals(CaseIndex.RELATIONSHIP_EXTENSION) &&
-                        node[0].equals(STATUS_LIVE)) {
+            int[] node = g.getNode(index);
```

Comment:
Since we don't care about the case ID stored in the node, I would change this to `int caseStatus = g.getNode(index)[0];`